app/services/protocol_importers/search_protocols_service.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module ProtocolImporters
+  class SearchProtocolsService
+    extend Service
+
+    attr_reader :errors, :protocols_list
+
+    CONSTANTS = Constants::PROTOCOLS_IO_V3_API
+
+    def initialize(protocol_source:, query_params: {})
+      @protocol_source = protocol_source
+      @query_params = query_params
+      @errors = Hash.new { |h, k| h[k] = {} }
+    end
+
+    def call
+      return self unless valid?
+
+      api_response = api_client.protocol_list(@query_params)
+
+      @protocols_list = normalizer.normalize_list(api_response)
+
+      self
+    end
+
+    def succeed?
+      @errors.none?
+    end
+
+    private
+
+    def valid?
+      # try if key is not empty
+      @errors[:invalid_params][:key] = 'Key cannot be empty' if @query_params[:key].blank?
+
+      # try if page id is ok
+      if @query_params[:page_id] && !@query_params[:page_id].positive?
+        @errors[:invalid_params][:page_id] = 'Page needs to be positive'
+      end
+
+      # try if order_field is ok
+      if CONSTANTS[:available_order_fields]&.exclude? @query_params[:order_field].to_sym
```

Comment:
This breaks generality. It should most likely be addressed in the API client, as well other checks in here. If that's the case, all these checks could go to API client and the `call` part could easily be done in controller since only two lines are needed for this service. Thoughts?

Reply:
I agree with you. Direct assigning constants breaks generality. I am thinking how much will be code 'inflated' when we add more than 1 endpoint. But not much if we handle params in ApiClient probably. And we can do mappings there if necessery.

Reply:
After discussion with @jbargu, we will stay with the service class for now. Error handling will be implemented there, validations check stays there for now and there is a fair possibility to execute this searching inside of background job in the future.